tests/0085-headers.cpp

```diff
@@ -0,0 +1,219 @@
+ public:
+  int seen_commit;
+  int seen_stats;
+
+  void offset_commit_cb (RdKafka::ErrorCode err,
+                         std::vector<RdKafka::TopicPartition*>&offsets) {
+    seen_commit++;
+    Test::Say("Got commit callback!\n");
+  }
+
+  void event_cb (RdKafka::Event &event) {
+    switch (event.type())
+      {
+      case RdKafka::Event::EVENT_STATS:
+        Test::Say("Got stats callback!\n");
+        seen_stats++;
+        break;
+      default:
+        break;
+    }
+  }
+};
+
+static void assert_all_headers_match(RdKafka::Headers *actual,
+                                     std::vector<RdKafka::Headers::Header> &expected) {
+  if (actual->size() != expected.size()) {
+    Test::Fail(tostr() << "Expected headers length to equal" 
+               << expected.size() << " instead equals " << actual->size() << "\n");
+  }
+
+  std::vector<RdKafka::Headers::Header> actual_headers = actual->get_all();
+  for(size_t i = 0; i < actual_headers.size(); i++) {
+    RdKafka::Headers::Header actual_header = actual_headers[i];
+    RdKafka::Headers::Header expected_header = expected[i];
+    std::string actual_key = actual_header.key();
+    std::string actual_value = std::string(actual_header.value());
+    std::string expected_key = expected_header.key();
+    std::string expected_value = std::string(expected_header.value());
+    if (actual_key != expected_key) {
+      Test::Fail(tostr() << "Header key does not match, expected '" 
+                 << actual_key << "' but got '" << expected_key << "'\n");
+    }
+    if (actual_value != expected_value) {
+      Test::Fail(tostr() << "Header value does not match, expected '" 
+                 << actual_value << "' but got '" << expected_value << "'\n");
+    }
+  }
+}
+
+static void test_n_headers (int n, const char* message) {
+  std::string topic = Test::mk_topic_name("0085-headers", 1);
+  RdKafka::Conf *conf;
+  std::string errstr;
+
+  Test::conf_init(&conf, NULL, 0);
+
+  Test::conf_set(conf, "group.id", topic);
+  Test::conf_set(conf, "group.id", topic);
+  Test::conf_set(conf, "socket.timeout.ms", "10000");
+  Test::conf_set(conf, "enable.auto.commit", "false");
+  Test::conf_set(conf, "enable.partition.eof", "false");
+  Test::conf_set(conf, "auto.offset.reset", "earliest");
+  Test::conf_set(conf, "statistics.interval.ms", "1000");
+
+  MyCbs cbs;
+  cbs.seen_commit = 0;
+  cbs.seen_stats  = 0;
+  if (conf->set("offset_commit_cb", (RdKafka::OffsetCommitCb *)&cbs, errstr) !=
+      RdKafka::Conf::CONF_OK)
+    Test::Fail("Failed to set commit callback: " + errstr);
+  if (conf->set("event_cb", (RdKafka::EventCb *)&cbs, errstr) !=
+      RdKafka::Conf::CONF_OK)
+    Test::Fail("Failed to set event callback: " + errstr);
+
+  RdKafka::Producer *p = RdKafka::Producer::create(conf, errstr);
+  if (!p)
+    Test::Fail("Failed to create Producer: " + errstr);
+
+  RdKafka::ErrorCode err;
+
+  std::vector<RdKafka::Headers::Header> headers_arr;
+  for (int i = 0; i < n; ++i) {
```

Comment:
Isn't it better to create the Headers object prior to this loop and appending directly to it instead of first constructing a vector to pass